simplex/storage.go

```diff
@@ -0,0 +1,269 @@
+		}
+		return nil, simplex.Finalization{}, false
+	}
+
+	finalization, found, err := s.retrieveFinalization(seq)
+	if !found || err != nil {
+		return nil, simplex.Finalization{}, false
+	}
+
+	vb := &Block{vmBlock: block, metadata: finalization.Finalization.ProtocolMetadata, blockTracker: s.blockTracker}
+	bytes, err := vb.Bytes()
+	if err != nil {
+		s.log.Error("failed to serialize block", zap.Error(err))
+		return nil, simplex.Finalization{}, false
+	}
+	vb.digest = computeDigest(bytes)
+
+	return vb, finalization, true
+}
+
+// Index indexes the finalization in the storage.
+// It stores the finalization bytes at the current height and increments the height.
+func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finalization simplex.Finalization) error {
+	bh := block.BlockHeader()
+	if bh.Seq == 0 {
+		s.log.Warn("attempted to index genesis block, which should not be indexed")
+		return errGenesisIndexed
+	}
+
+	currentHeight := s.height.Load()
+	if currentHeight != bh.Seq {
+		return fmt.Errorf("%w: expected %d, got %d", errUnexpectedSeq, currentHeight, bh.Seq)
+	}
+
+	if s.blockTracker.lastIndexed != bh.Prev {
+		s.log.Warn("attempted to index block with mismatched previous digest",
+			zap.String("expected", s.blockTracker.lastIndexed.String()),
+			zap.String("got", bh.Prev.String()))
+
+		return fmt.Errorf("%w: expected %s, got %s", errMismatchedPrevDigest, s.blockTracker.lastIndexed, bh.Prev)
+	}
+
+	if !bytes.Equal(bh.Digest[:], finalization.Finalization.Digest[:]) {
+		return fmt.Errorf("%w: expected %d, got %d", errMismatchedDigest, bh.Digest, finalization.Finalization.Digest)
+	}
+
+	if finalization.QC == nil {
+		return errInvalidQC
+	}
+
+	seqBuff := make([]byte, 8)
+	binary.BigEndian.PutUint64(seqBuff, currentHeight)
+
+	finalizationBytes := finalizationToBytes(finalization)
+	if err := s.db.Put(seqBuff, finalizationBytes); err != nil {
+		return fmt.Errorf("failed to store finalization: %w", err)
+	}
+
+	err := s.blockTracker.indexBlock(ctx, bh.Digest)
+	if err != nil {
+		return fmt.Errorf("failed to index block: %w", err)
+	}
+
+	s.height.Add(1) // only increment height after successful indexing
+	return nil
+}
+
+// getGenesisBlock returns the genesis block wrapped as a Block instance.
+func getGenesisBlock(ctx context.Context, config *Config, blockTracker *blockTracker) (*Block, error) {
+	genesis := &Block{
+		metadata:     genesisMetadata,
+		blockTracker: blockTracker,
+	}
+
+	snowmanGenesis, err := getBlockAtHeight(ctx, config.VM, 0)
+	if err != nil {
+		return nil, err
+	}
+	genesis.vmBlock = snowmanGenesis
+
+	// set the digest
+	bytes, err := genesis.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	genesis.digest = computeDigest(bytes)
+
+	return genesis, nil
+}
+
+// retrieveFinalization retrieves the finalization at [seq].
+// If the finalization is not found, it returns false.
+func (s *Storage) retrieveFinalization(seq uint64) (simplex.Finalization, bool, error) {
```

Comment:
The `bool` is never really used from this function. It is only handled as `if !found || err != nil {`, so I think we should remove the `bool` and just report the `error`.
```suggestion
func (s *Storage) retrieveFinalization(seq uint64) (simplex.Finalization, bool, error) {
```